### PR TITLE
[Numpy] arctan2, a new branch for pr.

### PR DESCRIPTION
**Pull request by tingying2020**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- Create a new branch and move code to it.

@haojin2

### CODE REVIEW

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 1]
```diff
@@ -198,5 +198,74 @@ MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_minimum_scalar)
 .set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Compute<cpu, mshadow_op::minimum>)
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_minimum_scalar"});
 
+inline bool Arctan2OpType(const nnvm::NodeAttrs& attrs,
+	                          std::vector<int>* in_attrs,
+	                          std::vector<int>* out_attrs){
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]:
```c++
inline bool Arctan2OpType(const nnvm::NodeAttrs& attrs,
	                  std::vector<int>* in_attrs,
	                  std::vector<int>* out_attrs){
```

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 2]
```diff
@@ -198,5 +198,74 @@ MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_minimum_scalar)
...
+     TYPE_ASSIGN_CHECK(*in_attrs, 0, out_attrs->at(0));
+     TYPE_ASSIGN_CHECK(*in_attrs, 1, out_attrs->at(0));
+     //check if it is float16, float32 or float64. If not, raise error
+     if(in_attrs->at(0) > 2){
+     // do not support int now.
+     std::ostringstream os;
+     os << "Do not support `int` as input.\n";
+     throw ::mxnet::op::InferTypeError(os.str(), 0);
+     }
+     return out_attrs->at(0) != -1;
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]: 2-space indentations, please fix for all your code

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 3]
```diff
@@ -198,5 +198,74 @@ MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_minimum_scalar)
...
+     TYPE_ASSIGN_CHECK(*out_attrs, 0, in_attrs->at(0));
+     TYPE_ASSIGN_CHECK(*out_attrs, 0, in_attrs->at(1));
+     TYPE_ASSIGN_CHECK(*in_attrs, 0, out_attrs->at(0));
+     TYPE_ASSIGN_CHECK(*in_attrs, 1, out_attrs->at(0));
+     //check if it is float16, float32 or float64. If not, raise error
+     if(in_attrs->at(0) > 2){
+     // do not support int now.
+     std::ostringstream os;
+     os << "Do not support `int` as input.\n";
+     throw ::mxnet::op::InferTypeError(os.str(), 0);
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]: indentation here?

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 4]
```diff
@@ -198,5 +198,74 @@ MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_minimum_scalar)
...
+
+MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_rarctan2_scalar)
+.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Compute<cpu, mshadow_op::rarctan2>)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_npi_rarctan2_scalar"});
+
+MXNET_OPERATOR_REGISTER_BINARY(_backward_npi_arctan2_scalar)
+.add_argument("scalar", "float", "scalar value")
+.set_attr_parser([](NodeAttrs *attrs) { attrs->parsed = std::stod(attrs->dict["scalar"]); })
+.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Backward<
+     cpu, mshadow_op::arctan2_grad>);
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]:
```c++
.set_attr<FCompute>("FCompute<cpu>",
                    BinaryScalarOp::Backward<cpu, mshadow_op::arctan2_grad>);
```

Comment (haojin2) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: Similar for the case below.

File: tests/python/gpu/test_operator_gpu.py [hunk 5]
```diff
@@ -2300,6 +2300,81 @@ def test_math():
...
+                                        [(4, 5), (4, 5)],  # normal case
+                                        [(3, 2), (3, 2)],  # tall matrices
+                                        [(), ()],  # scalar only
+                                        [(3, 0, 2), (3, 0, 2)],  # zero-dim
+                                        [(3, 4, 5), (4, 1)],  # trailing dim broadcasting
+                                        [(3, 4, 5), ()],  # scalar broadcasting
+                                        [(), (1, 2, 3)],  # scalar broadcasting
+                                        [(4, 3), (4, 1)],  # single broadcasting
+                                        [(3, 4, 5), (3, 1, 5)]  # single broadcasting in the middle
+                                        ]:
```
Comment (haojin2) [inline tests/python/gpu/test_operator_gpu.py]: Align lines:
```python
        for shape1, shape2 in [[(1,), (1,)],  # single elements
                               [(4, 5), (4, 5)],  # normal case
                               [(3, 2), (3, 2)],  # tall matrices
                               [(), ()],  # scalar only
                               [(3, 0, 2), (3, 0, 2)],  # zero-dim
                               [(3, 4, 5), (4, 1)],  # trailing dim broadcasting
                               [(3, 4, 5), ()],  # scalar broadcasting
                               [(), (1, 2, 3)],  # scalar broadcasting
                               [(4, 3), (4, 1)],  # single broadcasting
                               [(3, 4, 5), (3, 1, 5)]  # single broadcasting in the middle
                               ]:
```

File: tests/python/gpu/test_operator_gpu.py [hunk 6]
```diff
@@ -2300,6 +2300,81 @@ def test_math():
...
+                np_backward_2 = -1 * x11 / (x11 * x11 + x21 * x21)
+                np_backward_1 = dimReduce(np_backward_1, x11)
+                np_backward_2 = dimReduce(np_backward_2, x21)
+                assert_almost_equal(x1.grad.asnumpy(), np_backward_1, rtol=rtol, atol=atol)
+                assert_almost_equal(x2.grad.asnumpy(), np_backward_2, rtol=rtol, atol=atol)
+
+                mx_out = mx.np.arctan2(x1, x2)
+                np_out = np.arctan2(x1.asnumpy(), x2.asnumpy())
+                assert_almost_equal(mx_out.asnumpy(), np_out, rtol=rtol, atol=atol)
+
```
Comment (haojin2) [inline tests/python/gpu/test_operator_gpu.py]: one more blank line below

File: tests/python/gpu/test_operator_gpu.py [hunk 7]
```diff
@@ -2300,6 +2300,81 @@ def test_math():
             for op in ops:
                 run_math(op, shape, dtype, check_value=check_value)
 
+@with_seed()
```
Comment (haojin2) [inline tests/python/gpu/test_operator_gpu.py]: One more blank line above, 2 blank lines between functions.

File: src/operator/numpy/np_elemwise_broadcast_op.cu [hunk 8]
```diff
@@ -42,6 +42,12 @@ NNVM_REGISTER_OP(_npi_mod)
 NNVM_REGISTER_OP(_npi_power)
 .set_attr<FCompute>("FCompute<gpu>", BinaryBroadcastCompute<gpu, mshadow_op::power>);
 
+NNVM_REGISTER_OP(_npi_arctan2)
+.set_attr<FCompute>("FCompute<gpu>", BinaryBroadcastCompute<gpu, mshadow_op::arctan2>);
+
+NNVM_REGISTER_OP(_backward_npi_arctan2)
+.set_attr<FCompute>("FCompute<gpu>", BinaryBroadcastBackwardUseIn<gpu,mshadow_op::arctan2_grad,
+                                                                    mshadow_op::arctan2_rgrad>);
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cu]: one more blank line below.

File: src/operator/numpy/np_elemwise_broadcast_op.cc (hunk 1 above)
Comment (tingying2020) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: Fixed, thank you.

File: src/operator/numpy/np_elemwise_broadcast_op.cc (hunk 2 above)
Comment (tingying2020) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: Fixed, thank you.

File: src/operator/numpy/np_elemwise_broadcast_op.cc (hunk 3 above)
Comment (tingying2020) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: Fixed, thank you.

File: src/operator/numpy/np_elemwise_broadcast_op.cc (hunk 4 above)
Comment (tingying2020) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: Fixed, thank you.

File: tests/python/gpu/test_operator_gpu.py (hunk 5 above)
Comment (tingying2020) [reply tests/python/gpu/test_operator_gpu.py]: Fixed, thank you.

File: tests/python/gpu/test_operator_gpu.py (hunk 6 above)
Comment (tingying2020) [reply tests/python/gpu/test_operator_gpu.py]: Fixed, thank you.

File: tests/python/gpu/test_operator_gpu.py (hunk 7 above)
Comment (tingying2020) [reply tests/python/gpu/test_operator_gpu.py]: Fixed, thank you.

File: src/operator/numpy/np_elemwise_broadcast_op.cu (hunk 8 above)
Comment (tingying2020) [reply src/operator/numpy/np_elemwise_broadcast_op.cu]: Fixed, thank you.

**Pull request closed without merging**